runtime/src/iree/hal/drivers/cuda/tracing.c

```diff
@@ -169,98 +195,188 @@ void iree_hal_cuda_tracing_context_free(
 void iree_hal_cuda_tracing_context_collect(
     iree_hal_cuda_tracing_context_t* context) {
   if (!context) return;
-  if (context->query_tail == context->query_head) {
-    // No outstanding queries.
+  iree_slim_mutex_lock(&context->event_mutex);
+
+  // No outstanding queries
+  if (!context->event_submitted_list_head) {
+    iree_slim_mutex_unlock(&context->event_mutex);
     return;
   }
   IREE_TRACE_ZONE_BEGIN(z0);
 
-  while (context->query_tail != context->query_head) {
-    // Compute the contiguous range of queries ready to be read.
-    // If the ringbuffer wraps around we'll handle that in the next loop.
-    uint32_t try_query_count =
-        context->query_head < context->query_tail
-            ? context->query_capacity - context->query_tail
-            : context->query_head - context->query_tail;
-    IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, (int64_t)try_query_count);
-
-    // Scan and feed the times to tracy, stopping when we hit the first
-    // unavailable query.
-    uint32_t query_base = context->query_tail;
-    uint32_t read_query_count = 0;
-    for (uint32_t i = 0; i < try_query_count; ++i) {
-      // Ensure the event has completed; will return CUDA_ERROR_NOT_READY if
-      // recorded but not retired or any other deferred error.
-      uint16_t query_id = (uint16_t)(query_base + i);
-      CUevent query_event = context->event_pool[query_id];
-      CUresult result = context->symbols->cuEventQuery(query_event);
+  iree_hal_cuda_tracing_context_event_t* events =
+      context->event_submitted_list_head;
+  uint32_t read_query_count = 0;
+  while (events) {
+    iree_hal_cuda_tracing_context_event_t* event = events;
+    while (event) {
+      uint32_t query_id = (uint32_t)(event - &context->event_pool[0]);
+
+      CUresult result = context->symbols->cuEventQuery(event->event);
+      while (result == CUDA_ERROR_NOT_READY) {
```

Comment:
Hmm, this is a busy loop? Can we do better?

Reply:
We can call `cuEventSynchronize` to actively wait for the event (although it looks like it may just busy loop anyway). That being said I did this to be equivalent to the previous behavior (which would busy loop as long as result != CU_SUCCESS. (in pseudocode it looked like)
```
while there are events {
   try_query_count = contiguous_range_in_circular_array
   for i =  0 ... try_query_count  {
     if CUDA_SUCCESS != cuEventQuery(event);
        break
     }
}
```

Reply:
+1 for `cuEventSynchronize`.